### PR TITLE
WIP: Generic Matrix Multiplication

### PR DESCRIPTION
**Pull request by americast**
Naive matmul implemented. Tiling method is still in WIP, tests failing currently.
@SimonDanisch It would be great if you could kindly review this.

### CODE REVIEW

**Comment by SimonDanisch** [comment]
https://ci.appveyor.com/project/SimonDanisch/gpuarrays-jl/build/1.0.266/job/n2ervnv88u36dt4d#L33
Please make sure to not include CUDAnative ;)

**Comment by maleadt** [comment]
Huh, not sure why CI didn't went though. I've redelivered the payload manually.

**Comment by maleadt** [comment]
Did you manage to fix the `@cuprintf` corruption you mentioned before?

Also, I just came across https://cnugteren.github.io/tutorial/pages/page1.html, but I assume you already know about that. You might be interested in the experimental support for `__ldg` in CUDAnative for Julia 0.7, see https://github.com/JuliaGPU/CUDAnative.jl/pull/158

**Comment by americast** [comment]
@maleadt Not really. We were making use of `JLArray` for debugging stuff. Also, in place of `@cuprintf`, we made use of [`save_print`](https://github.com/JuliaPlots/Makie.jl/blob/sd/newdesign/src/logging.jl#L87) while testing with `JLArray` so that sync is not disturbed.

By the way, thanks for the resource. We were following it closely for our implementation. Experimental support for `__ldg` in Julia 0.7 sounds interesting!

**Comment by maleadt** [comment]
Well, file issues on CUDAnative then. I can't get this code to compile even, having tried 641bf3e7386a3362f14e26db70a15034a7e7acd8 (just before the `@cuprintf` removal), and reducing the kernel to something that compiles yields valid output... Same with the `@cuStaticShared(::UInt32)` issue @SimonDanisch hinted about on Slack, I'd love to help but at least file some issues with a MWE... I don't have time to reverse-engineer your WIP here.

**Comment by SimonDanisch** [comment]
the `@cuprintf` got resolved... @americast will open an issue with an MWE for `@cuStaticShared`

File: src/matmul.jl [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            return
+        end
+    end
+
+    # Local memory to fit a tile of TS*TS elements of A and B
+    Asub = @LocalMemory(state, T, IntTS²)
+    Bsub = @LocalMemory(state, T, IntTS²)
+
+    # Initialise the accumulation register
+    acc = zeros(T, WPT)
```
Comment (SimonDanisch) [inline src/matmul.jl]: zeros on the GPU?!

File: src/matmul.jl [hunk 2]
```diff
@@ -59,12 +59,14 @@ end
 function matmul!(dest::GPUArray, a::GPUArray{T, 2}, b::GPUArray{T, 2}) where T
     Asize = size(a)
     Bsize = size(b)
-    TS = UInt32(32)
+    device = GPUArrays.device(a)
+    thr = GPUArrays.threads(device)
+    TS = Int(sqrt(thr))
```
Comment (SimonDanisch) [inline src/matmul.jl]: needs to be `ceil(Int, sqrt(thr))` for correctness.. Although I suppose, max threads should always be powers of 2

File: src/matmul.jl [hunk 3]
```diff
@@ -59,12 +59,14 @@ end
...
     outSize = UInt32.(size(dest))
     Asize = UInt32.(Asize)
     Bsize = UInt32.(Bsize)
     config = ((div(Asize[1], TS), div(Bsize[2], TS)), (TS, TS))
-    gpu_call(matmul_kernel, dest, (a,b, dest, Asize, Bsize, outSize, Val{UInt32(32)}(), Val{UInt32(1024)}(), Val{1024}()), config)
+    gpu_call(matmul_kernel, dest, (a,b, dest, Asize, Bsize, outSize, Val{UInt32(TS)}(), Val{UInt32(TS^2)}(), Val{TS^2}()), config)
```
Comment (SimonDanisch) [inline src/matmul.jl]: can you just have the int one, and convert it in the kernel to UInt32?

Comment (americast) [reply src/matmul.jl]: Sure, but how does that help? :thinking:

Comment (maleadt) [reply src/matmul.jl]: SImplifies the kernel signature?
Also, just FYI @SimonDanisch , I've been moving away from UInt32 everywhere in CUDAnative; I haven't seen it improve performance and it just messes up readability (and hindered some LLVM-level optzns). I'm rather betting on using Cassette to do wholesale Int->Int32 transformations.

**Comment by maleadt** [comment]
This is pretty stale. Besides, we have a (naive) matmul now: https://github.com/JuliaGPU/GPUArrays.jl/pull/235

**Pull request closed without merging**